dask/blockwise.py

```diff
@@ -318,6 +336,25 @@ def __dask_distributed_pack__(self, client):
         # All blockwise tasks will depend on the futures in `indices`
         global_dependencies = tuple(stringify(f.key) for f in indices_unpacked_futures)
 
+        # Check if we need to serialize the elements of `io_deps`
+        # with `pickle` (i.e. something in the required inputs is
+        # not msgpack serializable)
+        if msgpack:
+            for _, input_map in self.io_deps.items():
+                check = True
+                for k in input_map:
+                    if check:
+                        # Test msgpack on the very first element.
+                        # If this fails, we will pickle everything.
+                        # Note that we may want to skip cases where
+                        # the element is something like `None`.
+                        try:
+                            msgpack.packb(input_map[k])
+                            break  # No need to pickle - Bail
+                        except TypeError:
+                            check = False
+                    input_map[k] = pickle.dumps(input_map[k])
```

Comment:
I would like to come up with a "nicer" solution that this, but I haven't had any bright ideas yet.  In general, we need to handle the case that the elements of `io_deps` are not `msgpack` serializable.  This happens, for example, when the parquet "pyarrow-dataset" engine needs to include a distinct `ParquetFileFragment` for each output partition.  The majority of cases should **not** require pickling.

The solution above is to check if `msgpack.packb` is successful for the first element of each collection in `io_deps`.  If the check succeeds, we simply skip serialization for that collections.  If the check fails, each element (for that collection) is pickled individually.  This covers the parquet case, because all partitions will include a `ParquetFileFragment` object if that specific code path is used.  Note that we must individually pickle each element (for each collection) of `io_deps`, because the scheduler needs to be able to select/isolate each element at graph materialization time (without calling `pickle.loads`).

An alternative solution might be to add an additional kwarg to specify if any of the collections in `io_deps` require special serializations.  We could also introduce a specialized `IODependencies` class for `io_deps` (instead of using a dict of dicts) which is capable of embedding this type of information.